### PR TITLE
Add tests and refactor

### PR DESCRIPTION
**Pull request by sophiebrown01**
This PR makes an initial attempt to add a broad covering of functional and model tests to the app.

The app has then been refactored on green, moving more of the functionality into models where appropriate.

This PR also includes `SurveyVisit` as a new concept, which means we can now accurately associate surveys to visits and get the real list of pages visited for a given survey, not _all the pages_ any given visitor has visited.

### CODE REVIEW

**Comment by sophiebrown01** [comment]
FYI I'll be fixing up the commit history.

File: spec/factories/organisations.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+FactoryBot.define do
+  factory :organisation do
+    name { "MyString" }
```
Comment (oscarwyatt) [inline spec/factories/organisations.rb]: There are a few of these things left as "MyString" in this commit, would be nice to have more plausible factory data (inevitably, someone will want to change this down the road and it'd be nice to have a clean commit history as much as possible). Perhaps the Ministry for Silly Walks will live on

File: app/models/page.rb [hunk 2]
```diff
@@ -1,14 +1,14 @@
-require 'elasticsearch/model'
+# require 'elasticsearch/model'
```
Comment (oscarwyatt) [inline app/models/page.rb]: Can remove this commit 
(https://github.com/alphagov/govuk-user-intent-survey-explorer/pull/32/commits/01b2dd4c8a93feeeb8aac46840b3cc8e47fedd77)

File: app/models/survey.rb [hunk 3]
```diff
@@ -1,4 +1,4 @@
-require 'elasticsearch/model'
+# require 'elasticsearch/model'
```
Comment (oscarwyatt) [inline app/models/survey.rb]: Can remove this commit 
(https://github.com/alphagov/govuk-user-intent-survey-explorer/pull/32/commits/01b2dd4c8a93feeeb8aac46840b3cc8e47fedd77)

File: spec/features/trending_spec.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
+require 'spec_helper'
+
+RSpec.feature 'trend page' do
+  # before(:each) do
```
Comment (oscarwyatt) [inline spec/features/trending_spec.rb]: This commented out code is revived in a later commit, would be nice to have it consolidated (I don't like having commented out code in a repo, I have written a short essay on the subject, available from any good bookseller)

File: spec/features/trending_spec.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+  # let(:top_page) { FactoryBot.create(:page, base_path: '/department-of-wizards') }
+  # let(:top_page_visit) { FactoryBot.create(:visit) }
+  # let(:page_visit) { FactoryBot.create(:page_visit, page: top_page) }
+
+  before(:each) do
+    @top_page = FactoryBot.create(:page, base_path: '/department-of-wizards')
+    @page_visit = FactoryBot.create(:page_visit, page: @top_page)
+
+    @phrase = FactoryBot.create(:phrase, phrase_text: 'abra cadabra')
```
Comment (oscarwyatt) [inline spec/features/trending_spec.rb]: Really pedantic in order to make these tests as readable/maintainable as possible we shouldn't really be defining properties of these objects in the tests (unless it's an integral part of that specific test) so either remove these or put them into the factory, most of these lines should be something like `@phrase = FactoryBot.create(:phrase)`

EDIT: I appear to have made the same comment more than once!

File: spec/features/mentions_spec.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
+require 'spec_helper'
+
+RSpec.feature 'mentions page' do
+
+  before :each do
+    visitor = FactoryBot.create(:visitor)
+    survey1 = FactoryBot.create(:survey, started_at: '2020-04-02 00:00:00', visitor: visitor)
+    survey2 = FactoryBot.create(:survey, started_at: '2020-04-03 00:00:00', visitor: visitor)
+
+    @phrase = FactoryBot.create(:phrase, phrase_text: 'how government works')
```
Comment (oscarwyatt) [inline spec/features/mentions_spec.rb]: Unless the phrase text explicitly needs to be different for a test, it should be set in the factory. Keeping it there DRYs stuff up as you've got the same thing in other spec files

File: spec/features/mentions_spec.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    @survey_answer1 = FactoryBot.create(:survey_answer, survey: survey1, answer: 'I want to understand how government works')
+    @survey_answer2 = FactoryBot.create(:survey_answer, survey: survey2, answer: 'How government works is important')
+
+    FactoryBot.create(:survey_phrase, phrase: @phrase, survey_answer: @survey_answer1)
+    FactoryBot.create(:survey_phrase, phrase: @phrase, survey_answer: @survey_answer2)
```
Comment (oscarwyatt) [inline spec/features/mentions_spec.rb]: FactoryBot allows you to create associations within the factory definition so this shouldn't be necessary, should be able to do something like
```
factory :survey_phrase do
  phrase_one
  survey_answer_one
end
```

File: spec/features/pages_visited_spec.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    @phrase = FactoryBot.create(:phrase, phrase_text: 'how government works')
+    @survey_answer1 = FactoryBot.create(:survey_answer, survey: survey1, answer: 'I want to understand how government works')
+    @survey_answer2 = FactoryBot.create(:survey_answer, survey: survey2, answer: 'How government works is important')
+
+    FactoryBot.create(:survey_phrase, phrase: @phrase, survey_answer: @survey_answer1)
+    FactoryBot.create(:survey_phrase, phrase: @phrase, survey_answer: @survey_answer2)
+
+    user_visit = FactoryBot.create(:visit, visitor: visitor)
+    pages = FactoryBot.create_list(:page, 20)
+    pages.each {|p| FactoryBot.create(:page_visit, page: p, visit: user_visit) }
```
Comment (oscarwyatt) [inline spec/features/pages_visited_spec.rb]: As above, this should be defined in a factory

File: spec/features/pages_visited_spec.rb [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+    visit pages_visited_path(@phrase)
+
+    expect(page).to have_link("/test-page-1")
+    expect(page).to have_link("/test-page-15")
+    expect(page).to have_content("Next page")
+  end
+end
```
Comment (oscarwyatt) [inline spec/features/pages_visited_spec.rb]: This (and plenty of other files) seem to be missing a newline at the end (I think the linter should have picked this up?). You can configure rubymine to do this automatically which might make life easier

File: spec/models/device_spec.rb [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+    context 'with populated database' do
+      before :each do
+        desktop_device = FactoryBot.create(:device, name: 'Desktop')
+        tablet_device = FactoryBot.create(:device, name: 'Tablet')
+        visitor = FactoryBot.create(:visitor)
+        survey = FactoryBot.create(:survey, visitor: visitor, started_at: '2020-03-20')
+        survey_answer = FactoryBot.create(:survey_answer, survey: survey)
+        @phrase = FactoryBot.create(:phrase)
+        FactoryBot.create(:survey_phrase, phrase: @phrase, survey_answer: survey_answer)
```
Comment (oscarwyatt) [inline spec/models/device_spec.rb]: I _think_ this is another duplication of factories, again would be nice to DRY it up

File: spec/models/survey_phrase_spec.rb [hunk 11]
```diff
@@ -1,5 +1,53 @@
...
+
+        result = SurveyPhrase.mentions_by_date_range_for_phrase(@phrase, start_date, end_date)
+
+        expected_result = [[start_date, 1], [Date.new(2020, 3, 11), 3], [Date.new(2020, 3, 12), 0], [end_date, 3]]
+        expect(result).to eq(expected_result)
+      end
+    end
+  end
+
+  def create_mentions_for_date(date_str, number_of_mentions)
```
Comment (oscarwyatt) [inline spec/models/survey_phrase_spec.rb]: Feel free to ignore this because I'm not super familiar with factorybot and best practices but if there is a chance it might be used elsewhere I think this might want to live in the factory file. If not, I'm 50/50 about whether it should

File: app/models/survey_phrase.rb [hunk 12]
```diff
@@ -20,6 +19,6 @@ def self.mentions_by_date_range_for_phrase(phrase, start_date, end_date)
       end
     end
 
-    mentions
+    mentions.sort
```
Comment (oscarwyatt) [inline app/models/survey_phrase.rb]: More out of curiosity than anything else, why this change? I expect there is a good reason but looking at the change I don't understand why it's necessary (should be functionally the same and probably faster for the db to do it). Maybe update the commit msg

File: config/routes.rb [hunk 13]
```diff
@@ -3,11 +3,15 @@
...
+  # get 'phrase/:id', to: "phrase#show", as: :phrase
+  resources :phrases do
+    member do
+      get 'usage'
+    end
+  end
 
-  get 'phrase/mentions/:id', to: "number_of_mentions#show", as: :number_of_mentions
-  resources :number_of_mentions, only: [:show]
+  # get 'phrase/mentions/:id', to: "number_of_mentions#show", as: :mentions
```
Comment (oscarwyatt) [inline config/routes.rb]: If this isn't used any more it should be removed

File: config/routes.rb [hunk 14]
```diff
@@ -19,6 +19,6 @@
   resources :surveys, only: [:show]
 
   resources :pages_visited, only: [ :show ]
-  resources :phrase_usage, only: [ :show ]
+  # resources :phrase_usage, only: [ :show ]
```
Comment (oscarwyatt) [inline config/routes.rb]: if not used, remove

File: app/models/survey.rb [hunk 15]
```diff
@@ -7,6 +7,7 @@ class Survey < ApplicationRecord
   has_many :questions, through: :visits
```
Comment (oscarwyatt) [inline app/models/survey.rb:7]: This is great, as is the commit message (explains everything 👍) but the commit msg should wrap at [~72 characters](https://gds-way.cloudapps.digital/standards/source-code.html#structure)

File: app/controllers/pages_visited_controller.rb [hunk 16]
```diff
@@ -7,7 +7,9 @@ def show
 private
 
   def unique_visitors_by_page(phrase)
-    Page.find_by_sql("select pages.id, pages.base_path, count(distinct(pv.visit_id)) as unique_visitors from mentions m join phrases on phrases.id = m.phrase_id join survey_answers sa on sa.id = m.survey_answer_id join surveys s on s.id = sa.survey_id join visits v on v.visitor_id = s.visitor_id join page_visits pv on pv.visit_id = v.id join pages on pages.id = pv.page_id where phrases.id = #{phrase.id} group by (pages.id) order by #{search_params[:sort_key]} #{search_params[:sort_direction]};")
+    # Page.find_by_sql("select pages.id, pages.base_path, count(distinct(pv.visit_id)) as unique_visitors from mentions m join phrases on phrases.id = m.phrase_id join survey_answers sa on sa.id = m.survey_answer_id join surveys s on s.id = sa.survey_id join visits v on v.visitor_id = s.visitor_id join page_visits pv on pv.visit_id = v.id join pages on pages.id = pv.page_id where phrases.id = #{phrase.id} group by (pages.id) order by #{search_params[:sort_key]} #{search_params[:sort_direction]};")
```
Comment (oscarwyatt) [inline app/controllers/pages_visited_controller.rb]: remove

File: app/models/page.rb [hunk 17]
```diff
@@ -11,4 +11,14 @@ class Page < ApplicationRecord
     indexes :base_path, type: "text", analyzer: "english" do
     end
   end
+
+  def self.unique_visitors_for_phrase(phrase, start_date, end_date, sort_key: "unique_visitors", sort_dir: "desc")
+    date_range = start_date..end_date
+
+    Page.joins(page_visits: [{ visit: [{ survey_visit: [{ survey: [{ survey_answers: [{ mentions: :phrase }] }] }] }] }])
```
Comment (oscarwyatt) [inline app/models/page.rb:38]: WOW! I have a feeling this could cause performance problems down the line but it's ok for the moment

File: spec/features/pages_visited_spec.rb [hunk 18]
```diff
@@ -3,24 +3,29 @@
...
 
     expect(page).to have_link("/test-page-1")
     expect(page).to have_link("/test-page-15")
     expect(page).to have_content("Next page")
   end
 end
+
+def create_pages_and_page_visits_for_visit(number_of_pages, visit)
+  pages = FactoryBot.create_list(:page, number_of_pages)
+  pages.each { |p| FactoryBot.create(:page_visit, page: p, visit: visit) }
```
Comment (oscarwyatt) [inline spec/features/pages_visited_spec.rb]: super pedantic but I don't like having single character variable names, it's clearer and easier for other people if we rename `p` to `page` and increases the line length by only 6 characters

File: spec/models/page_spec.rb [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+        start_date = Date.new(2020, 3, 18)
+        end_date = Date.new(2020, 3, 21)
+
+        result = Page.unique_visitors_for_phrase(@phrase, start_date, end_date, sort_key: "base_path")
+
+        expect(result).to eq([["/what-government-does", 2], ["/government-departments", 5], ["/about-government", 4]])
+      end
+
+      it "returns ordered page data for phrase by direction and default key when direction is specified and key is not" do
+        additional_page = FactoryBot.create(:page, base_path: "/find-help")
```
Comment (oscarwyatt) [inline spec/models/page_spec.rb:69]: this test and the one prior both have `additional_page`, if there is no need for them to have different base paths could they be defined as a factory and the code re-used?

File: app/controllers/pages_visited_controller.rb [hunk 20]
```diff
@@ -7,7 +7,9 @@ def show
 private
```
Comment (oscarwyatt) [inline app/controllers/pages_visited_controller.rb:7]: This commit msg also needs some newlines

File: app/controllers/trending_controller.rb [hunk 21]
```diff
@@ -9,7 +9,9 @@ def index
 private
 
   def top_pages
-    Page.find_by_sql("select pages.*, count(page_visits.visit_id) as total_pageviews, concat('https://www.gov.uk', pages.base_path) as govuk_link from pages join page_visits on page_visits.page_id = pages.id group by pages.id limit 10")
+    # Page.find_by_sql("select pages.*, count(page_visits.visit_id) as total_pageviews, concat('https://www.gov.uk', pages.base_path) as govuk_link from pages join page_visits on page_visits.page_id = pages.id group by pages.id limit 10")
```
Comment (oscarwyatt) [inline app/controllers/trending_controller.rb]: remove

**Review by oscarwyatt — CHANGES_REQUESTED** [review]
Sterling work! Mostly fairly pedantic style things, key takeaway is I'd like to see the tests DRYed up a bit (I have a book on clean code that says each test should be no more than 3 lines long, I think thats a but OTT but I think it's a good goal to make tests as short as possible)

File: spec/factories/organisations.rb (hunk 1 above)
Comment (sophiebrown01) [reply spec/factories/organisations.rb]: Fair point, but unless I'm testing the thing specifically then I'm not concerned with what FactoryBot populates properties with. This can be seen in tests where we specifically set model properties under test.

File: app/models/page.rb (hunk 2 above)
Comment (sophiebrown01) [reply app/models/page.rb]: This commit isn't in this branch.

File: app/models/survey.rb (hunk 3 above)
Comment (sophiebrown01) [reply app/models/survey.rb]: This commit isn't in this branch.

File: spec/features/trending_spec.rb (hunk 4 above)
Comment (sophiebrown01) [reply spec/features/trending_spec.rb]: Thanks, yep. Not intentionally committed as commented-out code.

File: spec/features/trending_spec.rb (hunk 5 above)
Comment (sophiebrown01) [reply spec/features/trending_spec.rb]: FactoryBot will create associations if you don't define them, but the problem we have is that the majority of our database joins span 4-5 tables and results in us needing to create quite a few of the models in order to correctly hook everything up to get the right results on a page. Happy to review what we're using just to bear this caveat in mind.

This particular test was the first I wrote for this app and there are a few model instances that could be removed. At a bare minimum we'd need page_visit, survey, survey_user_group, survey_answer and mention.

File: app/models/page.rb (hunk 17 above)
Comment (sophiebrown01) [reply app/models/page.rb:38]: We may have to possibly rethink our database schema, if we do find performance problems further down the road.

File: spec/features/pages_visited_spec.rb (hunk 18 above)
Comment (sophiebrown01) [reply spec/features/pages_visited_spec.rb]: Happy to rename it in this method. It's probably `p` after I extracted from the test itself, where if you specify `page` then you're referencing the Capybara `page` var instead.

File: spec/models/page_spec.rb (hunk 19 above)
Comment (sophiebrown01) [reply spec/models/page_spec.rb:69]: It already is defined as a factory `FactoryBot.define(:page)`. The reason they have different base paths here is because we're testing the ordering.

File: app/controllers/trending_controller.rb (hunk 21 above)
Comment (sophiebrown01) [reply app/controllers/trending_controller.rb]: I suspect you're reviewing commit by commit. This has been removed.

**Comment by sophiebrown01** [comment]
Thanks for the feedback @oscarwyatt . Where possible the tests could be DRYed up somewhat, but at the same time it is important to make sure the tests are fully descriptive of what's under test; it'd be good to see what others think of that test recommendation about 3 lines per test, it feels somewhat arbitrary to me.

I'll go back over and double check the test set-ups are defining only what needs to be defined.

File: spec/features/trending_spec.rb (hunk 4 above)
Comment (sophiebrown01) [reply spec/features/trending_spec.rb]: Likewise, this comment isn't in this branch.

File: app/controllers/pages_visited_controller.rb (hunk 20 above)
Comment (sophiebrown01) [reply app/controllers/pages_visited_controller.rb:7]: Commit message looks fine here - could you elaborate please?

File: spec/factories/organisations.rb (hunk 1 above)
Comment (sophiebrown01) [reply spec/factories/organisations.rb]: Updated

File: spec/features/trending_spec.rb (hunk 5 above)
Comment (oscarwyatt) [reply spec/features/trending_spec.rb]: This is a lot tidier now, thanks 👍

File: app/models/page.rb (hunk 17 above)
Comment (oscarwyatt) [reply app/models/page.rb:38]: yup, as we're still learning it's fine but a good thing to have in the back of our minds 💯

Comment (sophiebrown01) [reply app/models/page.rb:38]: Absolutely.

File: app/controllers/pages_visited_controller.rb (hunk 20 above)
Comment (oscarwyatt) [reply app/controllers/pages_visited_controller.rb:7]: These commits

 https://github.com/alphagov/govuk-user-intent-survey-explorer/pull/32/commits/9a119685fb7d606ba4410996e5fd552e228d8273

https://github.com/alphagov/govuk-user-intent-survey-explorer/pull/32/commits/20560169dd8eb78cfa7ff9d45dbd08d13e40d69c

the gds way says commit messages should be around 72 characters per line, both have the message all in one line (they're good messages, no need to change the content)

**Comment by oscarwyatt** [comment]
> Where possible the tests could be DRYed up somewhat, but at the same time it is important to make sure the tests are fully descriptive of what's under test; it'd be good to see what others think of that test recommendation about 3 lines per test, it feels somewhat arbitrary to me.

Yeah, it's a balancing act between being fully descriptive and not too long - I think the balance is about right now so good work.

Three lines is completely arbitrary - just what the author of the book thought- I've always thought three lines is too short and you end up with loads of helper functions which makes it harder to work with so I certainly don't try to get it down to 3 - I see it more a guideline rather than a rule and I think there is something in making tests as short as it can reasonably be without making them lose their descriptive power

**Comment by sophiebrown01** [comment]
Along those lines, I did consider putting the test data into seeds and then have seeds run as part of test set-up, but then you lose the context of what is being set-up and have to look-up in your tests where this data comes from to be able to fully understand them (as you would with fixtures). 

As you mention, as long as we keep the set-up to only what's under test then this is probably the best approach to keep understanding within tests.

**Review by oscarwyatt — APPROVED** [review]
